// _eslint-config-feng@1.4.0@eslint-config-feng/index.js
'use strict';

// 官网 http://eslint.org/
// 中文文档 http://eslint.cn/docs/user-guide/configuring

module.exports = {
    rules: {
        // 'off' 或 0 - 关闭规则
        // 'warn' 或 1 - 开启规则，使用警告级别的错误：warn (不会导致程序退出)
        // 'error' 或 2 - 开启规则，使用错误级别的错误：error (当被触发的时候，程序会退出)

        // Possible Errors
        // enforce “for” loop update clause moving the counter in the right direction.
        // 'for-direction': 2, //

        // disallow await inside of loops
        // 'no-await-in-loop': 2, //

        // disallow comparing against -0
        // 'no-compare-neg-zero': 2, // eslint:recommended

        // 禁止条件表达式中出现赋值操作符
        'no-cond-assign': 2, // eslint:recommended

        // 禁用 console
        // 'no-console': 2, // eslint:recommended

        // 禁止在条件中使用常量表达式
        // 'no-constant-condition': 2, // eslint:recommended

        // 禁止在正则表达式中使用控制字符
        // 'no-control-regex': 2, // eslint:recommended

        // 禁用 debugger
        // 'no-debugger': 2, // eslint:recommended

        // 禁止 function 定义中出现重名参数
        // 'no-dupe-args': 2, // eslint:recommended

        // 禁止对象字面量中出现重复的 key
        // 'no-dupe-keys': 2, // eslint:recommended

        // 禁止出现重复的 case 标签
        // 'no-duplicate-case': 2, // eslint:recommended

        // 禁止出现空语句块
        // 'no-empty': 2, // eslint:recommended

        // 禁止在正则表达式中使用空字符集
        // 'no-empty-character-class': 2, // eslint:recommended

        // 禁止对 catch 子句的参数重新赋值
        // 'no-ex-assign': 2, // eslint:recommended

        // 禁止不必要的布尔转换
        // 'no-extra-boolean-cast': 2, // eslint:recommended

        // 禁止不必要的括号
        // 'no-extra-parens': 2, // eslint:recommended

        // 禁止不必要的分号
        // 'no-extra-semi': 2, // eslint:recommended

        // 禁止对 function 声明重新赋值
        // 'no-func-assign': 2, // eslint:recommended

        // 禁止在嵌套的块中出现变量声明或 function 声明
        // 'no-inner-declarations': 2, // eslint:recommended

        // 禁止 RegExp 构造函数中存在无效的正则表达式字符串
        // 'no-invalid-regexp': 2, // eslint:recommended

        // 禁止在字符串和注释之外不规则的空白
        'no-irregular-whitespace': 2, // eslint:recommended

        // 禁止把全局对象作为函数调用
        // 'no-obj-calls': 2, // eslint:recommended

        // 禁止直接调用 Object.prototypes 的内置属性
        // 'no-prototype-builtins': 2, //

        // 禁止正则表达式字面量中出现多个空格
        // 'no-regex-spaces': 2, // eslint:recommended

        // 禁用稀疏数组
        // 'no-sparse-arrays': 2, // eslint:recommended

        // disallow template literal placeholder syntax in regular strings
        // 'no-template-curly-in-string': 2, //

        // 禁止出现令人困惑的多行表达式
        'no-unexpected-multiline': 2, // eslint:recommended

        // 禁止在return、throw、continue 和 break 语句之后出现不可达代码
        // 'no-unreachable': 2, // eslint:recommended

        // 禁止在 finally 语句块中出现控制流语句
        // 'no-unsafe-finally': 2, // eslint:recommended

        // disallow negating the left operand of relational operators
        // 'no-unsafe-negation': 2, // eslint:recommended

        // 要求使用 isNaN() 检查 NaN
        // 'use-isnan': 2, // eslint:recommended

        // 强制使用有效的 JSDoc 注释
        // 'valid-jsdoc': [2, { // 强制使用有效的 JSDoc 注释
        //     requireParamDescription: false, // 允许在参数标签缺少描述
        //     requireReturnDescription: false, // 允许在返回标签缺少描述
        //     requireReturn: false, // 当且仅当 函数或方法有一个 return 语句 (该选项不适用于构造函数)
        //     // prefer: { returns: 'return' }, // 强制使用一致的文档标签，这些标签使用一个对象进行指定，其属性的意思是要使用的是值而不是键 (例如，'return': 'returns' 意思是使用 @returns 而不是 @return)
        // }],

        // 强制 typeof 表达式与有效的字符串进行比较
        // 'valid-typeof': 2, // eslint:recommended

        // Best Practices
        // 强制 getter 和 setter 在对象中成对出现
        // 'accessor-pairs': 2, //

        // 强制数组方法的回调函数中有 return 语句
        // 'array-callback-return': 2, //

        // 强制把变量的使用限制在其定义的作用域范围内
        // 'block-scoped-var': 2, //

        // enforce that class methods utilize this
        // 'class-methods-use-this': 2, //

        // 指定程序中允许的最大环路复杂度
        // 'complexity': 2, //

        // 要求 return 语句要么总是指定返回的值，要么不指定
        // 'consistent-return': 2, //

        // 强制所有控制语句使用一致的括号风格 只允许单行省略大括号
        curly: [2, 'multi-line'], // eslint:recommended

        // 要求 switch 语句中有 default 分支
        // 'default-case': 2, //

        // 强制在点号之前和之后一致的换行
        // 'dot-location': 2, // eslint:recommended

        // 强制尽可能地使用点号
        // 'dot-notation': 2, // eslint:recommended

        // 要求使用 === 和 !==
        eqeqeq: [2, 'always'], // eslint:recommended

        // 要求 for-in 循环中有一个 if 语句
        'guard-for-in': [1], //

        // 禁用 alert、confirm 和 prompt
        // 'no-alert': 2, //

        // 禁用 arguments.caller 或 arguments.callee
        'no-caller': 2, //

        // 不允许在 case 子句中使用词法声明
        // 'no-case-declarations': 2, // eslint:recommended

        // 禁止除法操作符显式的出现在正则表达式开始的位置
        // 'no-div-regex': 2, //

        // 禁止 if 语句中 return 语句之后有 else 块
        // 'no-else-return': 2, // eslint:recommended

        // 禁止出现空函数
        // 'no-empty-function': 2, //

        // 禁止使用空解构模式
        // 'no-empty-pattern': 2, // eslint:recommended

        // 禁止在没有类型检查操作符的情况下与 null 进行比较
        // 'no-eq-null': 2, //

        // 禁用 eval()
        // 'no-eval': 2, //

        // 禁止扩展原生类型
        'no-extend-native': 2, //

        // 禁止不必要的 .bind() 调用
        'no-extra-bind': 2, // eslint:recommended

        // 禁用不必要的标签
        // 'no-extra-label': 2, // eslint:recommended

        // 禁止 case 语句落空
        // 'no-fallthrough': 2, // eslint:recommended

        // 禁止数字字面量中使用前导和末尾小数点
        // 'no-floating-decimal': 2, // eslint:recommended

        // disallow assignments to native objects or read-only global variables
        // 'no-global-assign': 2, // eslint:recommended

        // 禁止使用短符号进行类型转换
        // 'no-implicit-coercion': 2, // eslint:recommended

        // 禁止在全局范围内使用变量声明和 function 声明
        // 'no-implicit-globals': 2, //

        // 禁止使用类似 eval() 的方法
        // 'no-implied-eval': 2, //

        // 禁止 this 关键字出现在类和类对象之外
        'no-invalid-this': 1, //

        // 禁用 __iterator__ 属性
        // 'no-iterator': 2, //

        // 禁用标签语句
        'no-labels': [2], //

        // 禁用不必要的嵌套块
        // 'no-lone-blocks': 2, //

        // 禁止在循环中出现 function 声明和表达式
        // 'no-loop-func': 2, //

        // 禁用魔术数字
        // 'no-magic-numbers': 2, //

        // 禁止使用多个空格
        'no-multi-spaces': [
            1,
            {
                exceptions: {
                    VariableDeclarator: true,
                },
            },
        ],

        // 禁止使用多行字符串
        // 'no-multi-str': 2, //

        // disallow new operators outside of assignments or comparisons
        // 'no-new': 2, //

        // 禁止对 Function 对象使用 new 操作符
        // 'no-new-func': 2, //

        // 禁止对 String，Number 和 Boolean 使用 new 操作符
        'no-new-wrappers': 2, //

        // 禁用八进制字面量
        // 'no-octal': 2, // eslint:recommended

        // 禁止在字符串中使用八进制转义序列
        // 'no-octal-escape': 2, //

        // 禁止对 function 的参数进行重新赋值
        // 'no-param-reassign': 2, //

        // 禁用 __proto__ 属性
        // 'no-proto': 2, //

        // 禁止多次声明同一变量
        // 'no-redeclare': 2, // eslint:recommended

        // disallow certain properties on certain objects
        // 'no-restricted-properties': 2, //

        // 禁止在 return 语句中使用赋值语句
        // 'no-return-assign': 2, //

        // disallow unnecessary return await
        // 'no-return-await': 2, //

        // 禁止使用 javascript: url
        // 'no-script-url': 2, //

        // 禁止自我赋值
        'no-self-assign': [2], // eslint:recommended

        // 禁止自身比较
        // 'no-self-compare': 2, //

        // 禁用逗号操作符
        // 'no-sequences': 2, //

        // 禁止抛出异常字面量
        'no-throw-literal': [1], //

        // 禁用一成不变的循环条件
        // 'no-unmodified-loop-condition': 2, //

        // 禁止出现未使用过的表达式
        // 'no-unused-expressions': 2, //

        // 禁用出现未使用过的标
        // 'no-unused-labels': 2, // eslint:recommended

        // 禁止不必要的 .call() 和 .apply()
        // 'no-useless-call': 2, //

        // 禁止不必要的字符串字面量或模板字面量的连接
        // 'no-useless-concat': 2, //

        // 禁用不必要的转义字符
        // 'no-useless-escape': 2, // eslint:recommended

        // disallow redundant return statements
        'no-useless-return': [1], // eslint:recommended

        // 禁用 void 操作符
        // 'no-void': 2, //

        // 禁止在注释中使用特定的警告术语
        // 'no-warning-comments': 2, //

        // 禁用 with 语句
        'no-with': 2, //

        // require using Error objects as Promise rejection reasons
        // 'prefer-promise-reject-errors': 2, //

        // 强制在parseInt()使用基数参数
        // 'radix': 2, //

        // disallow async functions which have no await expression
        // 'require-await': 2, //

        // 要求所有的 var 声明出现在它们所在的作用域顶部
        // 'vars-on-top': 2, //

        // 要求 IIFE 使用括号括起来
        // 'wrap-iife': 2, // eslint:recommended

        // 要求或禁止 “Yoda” 条件
        // 'yoda': 2, // eslint:recommended

        // Strict Mode
        // 要求或禁止使用严格模式指令
        // 'strict': 2, // eslint:recommended

        // Variables
        // 要求或禁止 var 声明中的初始化
        // 'init-declarations': 2, //

        // 禁止 catch 子句的参数与外层作用域中的变量同名
        // 'no-catch-shadow': 2, //

        // 禁止删除变量
        // 'no-delete-var': 2, // eslint:recommended

        // 不允许标签与变量同名
        // 'no-label-var': 2, //

        // 禁用特定的全局变量
        // 'no-restricted-globals': 2, //

        // 禁止变量声明与外层作用域的变量同名
        'no-shadow': [
            1,
            {
                builtinGlobals: true,
                hoist: 'all',
                allow: ['e', 'err', 'error', 'key', 'value', 'data', 'i', 'j', 'k', '$', 'jQuery'],
            },
        ],

        // 禁止将标识符定义为受限的名字 不重写保留字
        'no-shadow-restricted-names': 2, //

        // 禁用未声明的变量，除非它们在 /*global */ 注释中被提到
        // 'no-undef': 2, // eslint:recommended

        // disallow initializing variables to undefined
        // 'no-undef-init': 2, // eslint:recommended

        // 禁止将 undefined 作为标识符
        // 'no-undefined': 2, //

        // 禁止出现未使用过的变量
        'no-unused-vars': [
            1,
            {
                vars: 'all',
                args: 'none',
            },
        ],

        // 禁止在变量定义之前使用它们
        'no-use-before-define': [
            2,
            {
                variables: true,
                functions: false,
                classes: false,
            },
        ],

        // Node.js and CommonJS
        // require return statements after callbacks
        // 'callback-return': 2, //

        // 要求 require() 出现在顶层模块作用域中
        // 'global-require': 2, //

        // 要求回调函数中有容错处理
        // 'handle-callback-err': 2, //

        //
        // 'no-buffer-constructor': 2, //

        // 禁止混合常规变量声明和 require 调用
        // 'no-mixed-requires': 2, //

        // 禁止调用 require 时使用 new 操作符
        // 'no-new-require': 2, //

        // 禁止对 __dirname 和 __filename 进行字符串连接
        // 'no-path-concat': 2, //

        // 禁用 process.env
        // 'no-process-env': 2, //

        // 禁用 process.exit()
        // 'no-process-exit': 2, //

        // 禁用通过 require 加载的指定模块
        // 'no-restricted-modules': 2, //

        // 禁用同步方法
        // 'no-sync': 2, //

        // Stylistic Issues
        // enforce linebreaks after opening and before closing array brackets
        'array-bracket-newline': [2, { multiline: true }], // eslint:recommended

        // 强制数组方括号中使用一致的空格
        // 'array-bracket-spacing': 2, // eslint:recommended

        // enforce line breaks after each array element
        // 'array-element-newline': 2, // eslint:recommended

        // 强制在单行代码块中使用一致的空格
        // 'block-spacing': 2, // eslint:recommended

        // 强制在代码块中使用一致的大括号风格
        'brace-style': [2, '1tbs'], // eslint:recommended

        // 强制使用骆驼拼写法命名约定
        // camelcase: [2, { properties: 'never' }], //

        // enforce or disallow capitalization of the first letter of a comment
        // 'capitalized-comments': 2, // eslint:recommended

        // 要求或禁止末尾逗号
        'comma-dangle': [1, 'always-multiline'], // eslint:recommended

        // 强制在逗号前后使用一致的空格
        'comma-spacing': 2, // eslint:recommended

        // 强制使用一致的逗号风格
        'comma-style': 2, // eslint:recommended

        // 强制在计算的属性的方括号中使用一致的空格
        'computed-property-spacing': 2, // eslint:recommended

        // 当获取当前执行环境的上下文时，强制使用一致的命名
        // 'consistent-this': 2, //

        // 要求或禁止文件末尾存在空行
        'eol-last': 2, // eslint:recommended

        // require or disallow spacing between function identifiers and their invocations
        'func-call-spacing': 2, // eslint:recommended

        // require function names to match the name of the variable or property to which they are assigned
        // 'func-name-matching': 2, //

        // 要求或禁止使用命名的 function 表达式
        // 'func-names': 2, //

        // 强制一致地使用 function 声明或表达式
        // 'func-style': 2, //

        // 禁用指定的标识符
        // 'id-blacklist': 2, //

        // 强制标识符的最小和最大长度
        // 'id-length': 2, //

        // 要求标识符匹配一个指定的正则表达式
        // 'id-match': 2, //

        // 强制使用一致的缩进
        indent: [2, 4, { SwitchCase: 1 }], // eslint:recommended

        // 强制在 JSX 属性中一致地使用双引号或单引号
        // 'jsx-quotes': 2, // eslint:recommended

        // 强制在对象字面量的属性中键和值之间使用一致的间距
        'key-spacing': 2, // eslint:recommended

        // 强制在关键字前后使用一致的空格
        'keyword-spacing': 2, // eslint:recommended

        // enforce position of line comments
        // 'line-comment-position': 2, //

        // 强制使用一致的换行风格 unix换行符
        'linebreak-style': [2, 'unix'], // eslint:recommended

        // 要求在注释周围有空行
        // 'lines-around-comment': 2, // eslint:recommended

        // 强制可嵌套的块的最大深度
        // 'max-depth': 2, //

        // 强制一行的最大长度
        'max-len': [2, 120],

        // 强制最大行数
        // 'max-lines': 2, //

        // 强制回调函数最大嵌套深度
        // 'max-nested-callbacks': 2, //

        // 强制函数定义中最多允许的参数数量
        'max-params': [
            1,
            {
                max: 5,
            },
        ],

        // 强制函数块最多允许的的语句数量
        // 'max-statements': 2, //

        // 强制每一行中所允许的最大语句数量
        // 'max-statements-per-line': 2, //

        // enforce newlines between operands of ternary expressions
        // 'multiline-ternary': 2, //

        // 要求构造函数首字母大写
        'new-cap': [2], //

        // 要求调用无参构造函数时有圆括号
        // 'new-parens': 2, // eslint:recommended

        // 要求方法链中每个调用都有一个换行符
        // 'newline-per-chained-call': 2, //

        // 禁用 Array 构造函数
        'no-array-constructor': 2, //

        // 禁用按位运算符
        // 'no-bitwise': 2, //

        // 禁用 continue 语句
        // 'no-continue': 2, //

        // 禁止在代码后使用内联注释
        // 'no-inline-comments': 2, //

        // 禁止 if 作为唯一的语句出现在 else 语句中
        // 'no-lonely-if': 2, // eslint:recommended

        // 禁止混合使用不同的操作符
        // 'no-mixed-operators': 2, //

        // 禁止空格和 tab 的混合缩进
        'no-mixed-spaces-and-tabs': 2, // eslint:recommended

        // disallow use of chained assignment expressions
        'no-multi-assign': [1], //

        // 禁止出现多行空行
        'no-multiple-empty-lines': [2, { max: 2 }], // eslint:recommended

        // 禁用否定的表达式
        // 'no-negated-condition': 2, //

        // 禁用嵌套的三元表达式
        'no-nested-ternary': [1], //

        // 禁用 Object 的构造函数
        'no-new-object': 2, //

        // 禁用一元操作符 ++ 和 --
        // 'no-plusplus': 2, //

        // 禁用特定的语法
        // 'no-restricted-syntax': 2, //

        // disallow all tabs
        'no-tabs': 2, //

        // 禁用三元操作符
        // 'no-ternary': 2, //

        // 禁用行尾空格
        'no-trailing-spaces': 2, // eslint:recommended

        // 禁止标识符中有悬空下划线
        // 'no-underscore-dangle': 2, //

        // 禁止可以在有更简单的可替代的表达式时使用三元操作符
        // 'no-unneeded-ternary': 2, // eslint:recommended

        // 禁止属性前有空白
        // 'no-whitespace-before-property': 2, // eslint:recommended

        // enforce the location of single-line statements
        // 'nonblock-statement-body-position': 2, // eslint:recommended

        // 强制大括号内换行符的一致性
        // 'object-curly-newline': 2, // eslint:recommended

        // 强制在大括号中使用一致的空格
        // 'object-curly-spacing': 2, // eslint:recommended

        // 强制将对象的属性放在不同的行上
        // 'object-property-newline': 2, // eslint:recommended

        // 强制函数中的变量要么一起声明要么分开声明
        'one-var': [
            2,
            {
                var: 'never',
                let: 'never',
                const: 'never',
            },
        ],

        // require or disallow newlines around variable declarations
        // 'one-var-declaration-per-line': 2, // eslint:recommended

        // 要求或禁止在可能的情况下使用简化的赋值操作符
        // 'operator-assignment': 2, // eslint:recommended

        // 强制操作符使用一致的换行符
        // 'operator-linebreak': 2, // eslint:recommended

        // 要求或禁止块内填充
        'padded-blocks': [2, 'never'], // eslint:recommended

        // require or disallow padding lines between statements
        // 'padding-line-between-statements': 2, // eslint:recommended

        // 要求对象字面量属性名称用引号括起来
        // 'quote-props': 2, // eslint:recommended

        // 强制使用一致的反勾号、双引号或单引号
        quotes: [2, 'single'], // eslint:recommended

        // 要求使用 JSDoc 注释
        // 'require-jsdoc': [1, {
        //     require: {
        //         FunctionDeclaration: true,
        //         MethodDefinition: true,
        //         ClassDeclaration: true,
        //     },
        // }],

        // require or disallow semicolons instead of ASI
        semi: 1, // eslint:recommended

        // 强制分号之前和之后使用一致的空格
        'semi-spacing': 2, // eslint:recommended

        // 强制分号的位置
        // 'semi-style': 2, // eslint:recommended

        // require object keys to be sorted
        // 'sort-keys': 2, //

        // 要求同一个声明块中的变量按顺序排列
        // 'sort-vars': 2, //

        // 强制在块之前使用一致的空格
        'space-before-blocks': 2, // eslint:recommended

        // 强制在 function的左括号之前使用一致的空格
        // 'space-before-function-paren': [2, 'never'], // eslint:recommended

        // 强制在圆括号内使用一致的空格
        // 'space-in-parens': 2, // eslint:recommended

        // 要求操作符周围有空格
        'space-infix-ops': 2, // eslint:recommended

        // 强制在一元操作符前后使用一致的空格
        'space-unary-ops': 2, // eslint:recommended

        // 强制在注释中 // 或 /* 使用一致的空格
        'spaced-comment': [2, 'always'], // eslint:recommended

        // enforce spacing around colons of switch statements
        // 'switch-colon-spacing': 2, // eslint:recommended

        // require or disallow spacing between template tags and their literals
        // 'template-tag-spacing': 2, // eslint:recommended

        // 要求或禁止 Unicode 字节顺序标记 (BOM)
        // 'unicode-bom': 2, // eslint:recommended

        // 要求正则表达式被括号括起来
        // 'wrap-regex': 2, // eslint:recommended

        // ECMAScript 6
        // 要求箭头函数体使用大括号
        // 'arrow-body-style': 2, // eslint:recommended

        // 要求箭头函数的参数使用圆括号
        'arrow-parens': [2, 'as-needed'], // eslint:recommended

        // 强制箭头函数的箭头前后使用一致的空格
        // 'arrow-spacing': 0, // eslint:recommended

        // 要求在构造函数中有 super() 的调用
        // 'constructor-super': 2, // eslint:recommended

        // 强制 generator 函数中 * 号周围使用一致的空格
        'generator-star-spacing': [2, 'after'], // eslint:recommended

        // 禁止修改类声明的变量
        // 'no-class-assign': 2, // eslint:recommended

        // disallow arrow functions where they could be confused with comparisons
        // 'no-confusing-arrow': 2, // eslint:recommended

        // 禁止修改 const 声明的变量
        // 'no-const-assign': 2, // eslint:recommended

        // 禁止类成员中出现重复的名称
        // 'no-dupe-class-members': 2, // eslint:recommended

        // disallow duplicate module imports
        // 'no-duplicate-imports': 2, //

        // disallow new operators with the Symbol object
        'no-new-symbol': 2, // eslint:recommended

        // disallow specified modules when loaded by import
        // 'no-restricted-imports': 2, //

        // 禁止在构造函数中，在调用 super() 之前使用 this 或 super
        'no-this-before-super': 2, // eslint:recommended

        // disallow unnecessary computed property keys in object literals
        // 'no-useless-computed-key': 2, // eslint:recommended

        // 禁用不必要的构造函数
        // 'no-useless-constructor': 2, //

        // disallow renaming import, export, and destructured assignments to the same name
        // 'no-useless-rename': 2, // eslint:recommended

        // 要求使用 let 或 const 而不是 var
        'no-var': 2, // eslint:recommended

        // 要求或禁止对象字面量中方法和属性使用简写语法
        // 'object-shorthand': 2, // eslint:recommended

        // 要求使用箭头函数作为回调
        // 'prefer-arrow-callback': 2, // eslint:recommended

        // 要求使用 const 声明那些声明后不再被修改的变量
        // 'prefer-const': 2, // eslint:recommended

        // require destructuring from arrays and/or objects
        // 'prefer-destructuring': 2, //

        // disallow parseInt() in favor of binary, octal, and hexadecimal literals
        // 'prefer-numeric-literals': 2, // eslint:recommended

        // require rest parameters instead of arguments
        'prefer-rest-params': 2, //

        // 要求使用扩展运算符而非 .apply()
        // 'prefer-spread': 2, // eslint:recommended

        // 要求使用模板字面量而非字符串连接
        // 'prefer-template': 2, // eslint:recommended

        // 要求 generator 函数内有 yield
        // 'require-yield': 2, // eslint:recommended

        // enforce spacing between rest and spread operators and their expressions
        'rest-spread-spacing': 2, // eslint:recommended

        // 强制模块内的 import 排序
        // 'sort-imports': 2, // eslint:recommended

        // require symbol descriptions
        // 'symbol-description': 2, //

        // 要求或禁止模板字符串中的嵌入表达式周围空格的使用
        // 'template-curly-spacing': 2, // eslint:recommended

        // 强制在 yield* 表达式中 * 周围使用空格
        'yield-star-spacing': [2, 'after'], // eslint:recommended
    },
};
